### ISSUE TITLE
Regression with joins

### ISSUE THREAD

**Issue by lisinge** [slick/slick#1408]
I found another regression with joins and for comprehensions which is not solved by the solution in #1316. I added some test cases instead of plottering this issue.

1. Accessing foreign key relation through a many-to-many raises query compilation error as seen in this commit: lisinge/slick@da8dd9f580bbdff6d20472c2cf804ddd5adbf257
`a` has and belongs to many `b` through `c` and `d` belongs to `b`
a <-> b -> d

2. Sorting on relation in many-to-many for comprehension raises query compilation error as seen in this commit: lisinge/slick@4bf7e214ac347ef6ea6bcf034ac81c94a02c9578
`a` has and belongs to many `b` through `c`
a <-> b

**Comment by lisinge** [comment]
This is still a problem even after #1316 has been fixed

**Comment by lisinge** [comment]
@szeiger Have you had the chance to take a look at this?

**Comment by papauschek** [comment]
Yes, we're still experiencing the same problem with 3.1.1.

**Comment by deusaquilus** [comment]
(+1) Problem still exists in 3.1.1 today

**Comment by lisinge** [comment]
Ping @szeiger @cvogt, have you had the chance to look at this yet?

**Comment by cvogt** [comment]
@lisinge could you turn this into a PR that allows us to see the error in CI ?

**Comment by lisinge** [comment]
@cvogt i created the prs for you, they were in two different branches so i created #1512 and #1513

**Comment by deusaquilus** [comment]
Any progress on these? I'm stuck on 3.0.0 until they're fixed :(

**Comment by cvogt** [comment]
@radsaggi maybe you have time to look at these?
@szeiger is the expert in the area though

**Comment by deusaquilus** [comment]
I'm pretty sure this is a regression issue, it was working in 3.0.0. Any chance it could get fixed by 3.2.0?

**Comment by sweepy84** [comment]
+1 This is causing all sorts of issues for me. Looks like this isn't assigned to anyone... is it even been looked at?

**Comment by deusaquilus** [comment]
I've complained about this in multiple places a few times already but so far nothing has happened in months. Unfortunately I don't know enough about the ASTs to fix it myself.

**Comment by lisinge** [comment]
Yeah, i created this issue in december 2015 so it's been a while now.

I would love to try and help with fixing this issue too but my knowledge with the ASTs are also limited, i did what i could at least and added tests that reproduce the error.

**Comment by sweepy84** [comment]
Anybody looking at this??? Quite ridiculous... core functionality not working. Maybe stop working on improvements and fix major bugs? or am I being silly in thinking this is a major bug?? Can't access foreign key relation in a for comprehension!

**Comment by maximskripnik** [comment]
Shall we see solution in 3.2.x? Have to write plain sql queries cause of this.(

**Comment by lisinge** [comment]
@cvogt @szeiger Could we get an update on this issue?

Any info would be great, have you had a look at my test cases, are they of any help? Can i do something more to help?

**Comment by thomastoye** [comment]
Ran in to this issue while evaluating Slick nearly a year ago.. Crazy it's still not fixed, I'm not touching Slick again until it is, my usecase involved too much hassle working around this.

**Comment by cvogt** [comment]
@szeiger you got some time to look into this?

**Comment by deusaquilus** [comment]
Finally decided to look through the AST phases as this thing goes through, I'm pretty sure the issue is in RewriteJoins. There are cases for Bind -> Bind -> Pure, Bind -> Bind -> Filter, Bind -> Bind -> Join, Bind -> Bind -> Any Node, and Bind -> Pure but this kind of thing is producing a Bind -> Bind -> Bind which I don't think the expansions cover property (i.e there's a Bind node in the middle of the AST even after RewriteJoins is over). Could the answer be to write another case in RewriteJoins for this particular variation of the AST? Of course that begs the question of how this thing worked before RewriteJoins was introduced?

**Comment by deusaquilus** [comment]
Okay, been looking at this for while and have a possible solution, at least to  #1408 (i.e. the Joining issue).  The problem is that in the AST, references are propagated up the tree until a Bind node's 'select' clause can find the reference to the item in the ancestry of it's 'from' clause node. Using this sort of Bind(from:[...], select:Bind(from:[...],  select:Bind(...))) construct, multiple monadic references can be chained. The trouble is that in this particular case, filters in the B,C areas cannot reach the A reference in order to filter by the A id column (this symbol is s2 in the testManyToManyJoin hence the failure message).

Pease refer to #1673 where I have added new compiler phase that basically changes the AST behind something like this:

```scala
 def q1 = for {
   a <- as
   b <- a.bs
   d <- b.d
 } yield (a, b.id, d)
```

which is really this:
```scala
def q1 = for {
   a <- as
   b <- cs.filter(_.aId === a.id).flatMap(_.b)
   d <- b.d
 } yield (a, b.id, d)
```
Into an AST that results from this:
```scala
 def q1 = for {
   a <- as
   c <- cs.filter(_.aId === a.id)
   b <- c.b
   d <- b.d
 } yield (a, b.id, d)
```

This causes the original testManyToManyJoin test written by @lisinge to pass as well as a new test I have written called testManyToManyJoinTwice which looks at a double many-to-many case.

Unfortunately fixing the sortBy case is a bit harder because getting it work requires pulling up the SortBy node all the way to the top of the AST, I will work on that as I am able. Some relevant notes that I put together are also attached, I hope they are of use to anyone attempting to take up issues with the AST.

[Bind rewrite references.pdf](https://github.com/slick/slick/files/673712/Bind.rewrite.references.pdf)
[Rethinking bind rewrite.pdf](https://github.com/slick/slick/files/673713/Rethinking.bind.rewrite.pdf)

.

**Comment by deusaquilus** [comment]
Also, please note that the sortBy error situation is not limited to a many-to-many case, it occurs when **any** flat-mapped clause is followed by a sortBy as can be evidenced by the following test:

```scala
    class A(tag: Tag) extends Table[Int](tag, "a_simpleflatmapjoin") {
      def id = column[Int]("id", O.PrimaryKey)
      def * = id
    }
    lazy val as = TableQuery[A]
    class C(tag: Tag) extends Table[(Int, Int)](tag, "c_simpleflatmapjoin") {
      def id = column[Int]("id")
      def aId = column[Int]("aId")
      def * = (id, aId)
      def a = foreignKey("a_fk", aId, as)(_.id)
    }
    lazy val cs = TableQuery[C]

      def q1 = for {
        c <- cs
        a <-  c.a.sortBy(_.id)
      } yield (a, c)

    DBIO.seq(
      (as.schema ++ cs.schema).create,
      as ++= Seq(1),
      cs ++= Seq((2,1)),
      q1.result.named("q1").map(_.toSet shouldBe Set((1, 2)))
    )
```

This fails with the following error:
```
[error] Test com.typesafe.slick.testkit.tests.ThreeWayJoinTest.testManyToManyJoinSort[heap] failed: slick.SlickException: No type for symbol s26 found for Ref s26, took 0.799 sec
[error]     at slick.ast.Ref.withInferredType(Node.scala:558)
[error]     at slick.ast.Ref.withInferredType(Node.scala:551)
[error]     at slick.ast.Node$class.infer(Node.scala:92)
[error]     at slick.ast.Ref.infer(Node.scala:551)
[error]     at slick.ast.SimplyTypedNode$$anonfun$2.apply(Node.scala:128)
[error]     at slick.ast.SimplyTypedNode$$anonfun$2.apply(Node.scala:128)
[error]     at slick.ast.UnaryNode$class.mapChildren(Node.scala:229)
```

In this case I'm not sure what is wrong with the parse tree. For some reason, references do not propagate across a SortBy node like they would across a Bind node.

```
*** (slick.compiler.QueryCompiler) After phase rewriteJoins:
┃ Bind : Vector[t19<{s23: Int, s24: Int, s25: Int}>]
┃ ┣ from s28: Join Inner : Vector[(@t14<{id: Int, aId: Int}>, @t16<{id: Int}>)]
┃ ┃ ┣ left s26: Table c_simpleflatmapjoin : Vector[@t14<{id: Int, aId: Int}>]
┃ ┃ ┣ right s27: SortBy Ordering(Asc,NullsDefault) : Vector[@t16<{id: Int}>]
┃ ┃ ┃ ┣ from s12: Filter s10 : Vector[@t16<{id: Int}>]
┃ ┃ ┃ ┃ ┣ from s10: Table a_simpleflatmapjoin : Vector[@t16<{id: Int}>]
┃ ┃ ┃ ┃ ┗ where: Apply Function = : Boolean
┃ ┃ ┃ ┃   ┣ 0: Path s10.id : Int
┃ ┃ ┃ ┃   ┗ 1: Path s26.aId : Int
┃ ┃ ┃ ┗ by0: Path s12.id : Int
┃ ┃ ┗ on: LiteralNode true (volatileHint=false) : Boolean
┃ ┗ select: Pure t19 : Vector[t19<{s23: Int, s24: Int, s25: Int}>]
┃   ┗ value: StructNode : {s23: Int, s24: Int, s25: Int}
┃     ┣ s23: Path s28._2.id : Int
┃     ┣ s24: Path s28._1.id : Int
┃     ┗ s25: Path s28._1.aId : Int
```

In any case, the sortBy problem is a completely different issue and needs to be addressed by a different fix.

**Comment by lisinge** [comment]
Wow @deusaquilus, great work. Unfortunately i'm not anywhere close to as comfortable with the AST as you seem to have gotten, so i can't really help more than what i've already done with the tests.

Hopefully someone else can help you getting your fix merged and help with the sortBy fix.

**Comment by asheshambasta** [comment]
+1 still a problem. Surprising something as fundamental as this hasn't been fixed or hasn't seen a quicker resolution. Showstopper for us, I'm afraid.

**Comment by deusaquilus** [comment]
So first #1674 was on 3.1.2, then on 3.2.0, then on 3.2.1 on a two month old contribution to a Regression Issue which has been around for a year!? Meanwhile, people are lining up, asking, pleading, and begging for a fix for this issue! I spent two weeks digging through the query compiler trying to figure out how it fits together and nobody that knows what they're doing can even pipe up and give me two sentences of guidance!?

**Comment by sweepy84** [comment]
@deusaquilus it's just beyond ridiculous now...Typesafe/Lightbend have seriously dropped the ball on this! I think twice before chosing Scala/Slick for my projects now. I just can't afford wasting weeks to refactor an entire project because of their failure to catch/address such an intrusive bug... Slick is great, but as with any tech, I won't bother using it if it's not reliable.

What's worse than a bug? Not being able to rely on the developer's commitment to fix.

**Comment by szeiger** [comment]
Fixing the deadlock problem was the priority for 3.2. Now that this is done I will get back to the other open PRs for 3.2.1/3.3.

**Comment by hepin1989** [comment]
Previously ,@szeiger is a bitter busy on Scala 2.12,and small team,big project.I think we should understand and be easy on that.like @deusaquilus  and other contributors who spent much precious time on the digging,bug fixing,testing and documents ,the community owe you.

I think we could try to polish the PR more great and easy to merge,well tested and still try to push the PR ,like ping the members when you think it's good to go.

I know many guys in China love slicks very much,but them just could not found much time to fix the problem.

Anyway ,slick is not dying,just growing a little bit slower.

**Comment by cvogt** [comment]
@sweepy84 @deusaquilus Sorry about this guys. We are aware that these things weren't delivered very well and in active discussion how to improve things going forward. One option on the table is bringing in another paid person to help with more timely bug fixes, releases and supporting the community. We are currently figuring out if/how we can allocate the necessary funding for that.

**Comment by nicolasblaye** [comment]
Hey guys ! Can you tell update us on this issue? Encountered it in my project too.

**Comment by hvesalai** [comment]
@deusaquilus I'm collecting PR's for the next slick release (3.3). Did you have any further fixes you wanted to include related to this?

**Comment by ssdeep** [comment]
Has there been any resolution to this?  Especially the sortBy case?

**Comment by laurita** [comment]
I am having the same issue with `sortBy` (Slick version 3.3.0). Any update on this?
